### ISSUE TITLE
Question : pebblejs smoothness

### ISSUE THREAD

**Issue by eswak** [pebble/pebblejs#30]
Hi,

First of all, great job building this SDK ! I really enjoyed reading the docs.

I would like to develop apps for Pebble smartwatches, but as node.js developer who have not used C for years, the language barrier is quite a challenge for me (and the `snake_case` too, lol)... Pebblejs seems to be the solution.

I read the [online doc](http://developer.getpebble.com/docs/pebblejs) and one of the first thing written is : 
> Please be aware that as a result of Bluetooth round-trips for all actions, Pebble.js apps will use more power and respond slower to user interaction than a similar native app.

And I really did notice significant delays when running the provided default app. Sometimes, when I click a physical button on my watch, the action can take seconds to be effective.

So, before I start, I'd like to know a few things :
- Is it just that my phone's Bluetooth sucks ? (I don't think so, because I encounter the same problem on the CloudPebble emulator)
- Is there any chance the responsiveness of on-phone javascript pebble apps will get better in the future ? (i.e. shorter delays)
- Is there a way to "compile" a javascript application to make it a native pebble app ?
- Can I at least display a "loading" visual clue between user physical button and the end of the Bluetooth round-trip ?

Also, I plan to have a charting feature in my app (the lack of `line/path` `UI.Element` is not a problem, I'll just make bar charts for now), so I'd like to know if complex `UI.Window`s (with lots of `UI.Element`s in it) take significantly longer to be displayed ?

Thanks for reading :-)

Eswak

**Comment by Meiguro** [comment]
No, thank you for compliments on the doc! There are certainly rough parts, but sections are being improved as pebble.js is being built.

  * Now for the real topic: Bluetooth. I appreciate the info about the emulator, that's something new to me. In other issues in this repo, I have hinted that there may be something in App Message that is adding extra delays on top of Bluetooth. Experiencing delays in the emulator makes that more of a possibility. I will have to investigate that as soon as I find the time.
  * Compiling JavaScript applications is something I am looking into, but I can't make any promises. Investigating the above issue or this are the two primary ways latency can be reduced.
  * A loading visual cue is an interesting idea, I'll take that as a feature request for the action bar.
  * You are correct, pebble.js completely depends on the phone. It'll display a (currently ugly) notice to run the phone app.

Many elements can cause it to take longer to load, but not significantly so. You would need to be sending more than 1.5KB in order to start seeing delays due to excessive elements. Showing a UI.Rect is 29 bytes, complete with size and color. I do plan to eventually add GPath support so that you may draw lines however!

Latency is a huge issue however. The strongest use case for pebble.js is displaying lists of rapidly updating data from the internet, such as the latests posts from a popular source since that latency would be unavoidable even with a C app. But otherwise, if latency is a real problem. you can use pebble.js as a way to prototype your idea. The Pebble developers blog has been releasing C tutorials. You can prototype your app in pebble.js and write the same app in C afterwards.

**Comment by eswak** [comment]
Thanks a lot for all these responses :-)

I guess I'll start to read the C tutorials while prototyping in JS... :+1: 

Cheers !

**Comment by eswak** [comment]
this is completely off-topic but I just happened to back the Pebble Time Kickstarter and look what I saw on the confirmation screen :-p

![http://i.imgur.com/RbLJJ4b.png](http://i.imgur.com/RbLJJ4b.png)

**Comment by Meiguro** [comment]
Haha, I have to have my Pebble Time!

Interestingly, it seems the list of some other backers isn't very random. Most of those people were among the first to back the project. I wonder if it works by selecting a random slice/window from the chronological list, if it just displays the early backers for everyone, or the worst case, if it sorts using a comparator that just returns `random()`.

**Comment by andrei-markeev** [comment]
Hi guys!

First of all, Meiguro, huge thanks for your work, awesome project! :+1:

JS is gradually becoming the programming language of the world over last years, and having ability to develop in JS for Pebble is priceless. Thank you again!

Well, of course nothing is perfect and here I think I fully back the eswak on the idea of making PebbleJS somehow work standalone unless it explicitly needs the phone access. Myself I am very interested in caching scenarios, since even 2K storage can quite easily store e.g. a dozen of notes or calendar events, which can be synced only rarely thus significantly prolonging the watch battery life. And the battery life is one of the best Pebble features...!

Are there any news or plans in this direction? I think transpiling JS to C is actually quite realistic, although of course involves a lot of work...
